### PR TITLE
feat(history-logging): add history logging and use in feature flags

### PR DESCRIPTION
**Pull request by pauldambra**
For #8545 this adds an HistoricalVersion model and uses it in feature flags to save versions and expose them on the API

👉 *Stay up-to-date with our [coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

by adding tests and using it in context in #8570

### CODE REVIEW

File: posthog/models/historical_version.py
Comment (mariusandra) [inline posthog/models/historical_version.py:28]: Not sure I understand this line

File: posthog/api/test/test_feature_flag.py
Comment (mariusandra) [inline posthog/api/test/test_feature_flag.py:381]: How come this is false? Do we store the state before the changes... or after the changes?

**Review by mariusandra — COMMENTED** [review]
partial review, submitting because call coming up

File: posthog/api/test/test_feature_flag.py
Comment (pauldambra) [reply posthog/api/test/test_feature_flag.py:381]: This is testing the delete API call

That property is when we carry out a delete in the UI (which is a soft delete). So this flag hasn't been soft deleted and it's delete property is false but it has been deleted 🤯

File: posthog/models/historical_version.py
Comment (pauldambra) [reply posthog/models/historical_version.py:28]: Equivalent to pass by reference or value. If I hold a reference to X someone can change X and when I generate the log before and after that change the log is different but should be immutable

You need the state of the item as it was each time the log is written to. Not the state of the item as it is when the log is read from

You _could_ store the changes instead of the state and then you wouldn't care, but changing how you calculate changes would then also include changes over the data. This way if we alter how we calculate changes the log reflects the new logic when it is next read

File: posthog/api/test/test_feature_flag.py
Comment (pauldambra) [reply posthog/api/test/test_feature_flag.py:381]: I'll add a comment here to help the future traveller

File: posthog/models/historical_version.py
Comment (mariusandra) [reply posthog/models/historical_version.py:28]: This was my way of saying "the line is confusing or misleading, can you clear it up". E.g. why "should hold", not "actually does hold"? Why would I even assume that when I read a history log I get current data (as it is when I read it...?)? What is even "it" here, the log or the item?

I'm still not sure I understand this after the expanded explanation 😅

File: posthog/api/test/test_feature_flag.py
Comment (mariusandra) [reply posthog/api/test/test_feature_flag.py:389]: You added a comment 40 lines above here, I'd nit and add another `//` here as well, as we can't be sure future traveller reads from top to bottom...

**Review by mariusandra — COMMENTED** [review]
I'll admit that I'm running into the edge of my django knowledge again, and can't review with confidence. I can't see what downstream effect those changes in these mixins and models have, sorry.

This feature brings with it a lot of complexity that feels good to get right from the start, so would love a second opinion from someone. @macobo @Twixes  ?

File: posthog/models/historical_version.py
Comment (pauldambra) [reply posthog/models/historical_version.py:86]: I can't remember why I added this, and I don't think it's correct without an item_id anyway. I've removed it for now, it's cheap to add an index if we need it in future.

File: posthog/models/historical_version.py
Comment (pauldambra) [reply posthog/models/historical_version.py]: third time is the charm?

File: posthog/models/history_logging.py
Comment (pauldambra) [reply posthog/models/history_logging.py:66]: Given we have an unknown number of models this might be applied to 

(someone may only update to a version with a history log after we've added many types to track)

It seemed safer to add the import on demand than risk a migration that takes a long time

File: posthog/models/historical_version.py
Comment (Twixes) [inline posthog/models/historical_version.py:99]: The convention in Django is to do DB operations via `model.objects` (`objects` being the model's manager). For an example you can see how the `Organization` model has a custom `objects` attribute, which is a `class OrganizationManager(models.Manager)` instance.

File: posthog/models/history_logging.py
Comment (Twixes) [inline posthog/models/history_logging.py:72]: Who is History Hog? 🤔

File: posthog/models/historical_version.py
Comment (Twixes) [inline posthog/models/historical_version.py:56]: Will we be able to restore state from this?

File: posthog/api/test/test_feature_flag.py
Comment (hazzadous) [inline posthog/api/test/test_feature_flag.py:158]: To reduce dependencies on internal details, and make it easier to refactor, can we use the history http api instead of querying the db directly?

File: posthog/api/test/test_feature_flag.py
Comment (hazzadous) [inline posthog/api/test/test_feature_flag.py:433]: Can we consider that history would look like in cases of schema evolution?

e.g. 1. add a new field for feature flags "multivariate_type" with a default 2. feature flag updated 3. view history. Should / would multivariate_type be included in our history even if unchanged?

File: posthog/mixins.py
Comment (hazzadous) [inline posthog/mixins.py:37]: Can we specifically add to `FeatureFlagViewSet` rather than using a Mixin?

I don't see much benefit doing this, but makes the code non-obvious. Ideally I'd love to be able to look at FeatureFlagViewSet and see action create, update, destroy. This appears to be premature abstraction imo

File: posthog/models/historical_version.py
Comment (hazzadous) [inline posthog/models/historical_version.py:102]: Can we just add this to `FeatureFlagViewSet` directly? We can always factor out when we include other object types.

File: posthog/api/feature_flag.py
Comment (pauldambra) [reply posthog/api/feature_flag.py:205]: It was on the mixin initially and I ended up ignoring types. This way around I can inject specific types and avoid any dark magic

May well be that once there are several examples it makes sense to move this. But I'd favour being explicit now

File: posthog/models/historical_version.py
Comment (hazzadous) [inline posthog/models/historical_version.py:23]: Any estimate on the size of this table? I assume there are actually very few feature flags, and not that many of the other objects that we will be included?

File: posthog/api/test/test_feature_flag.py
Comment (hazzadous) [inline posthog/api/test/test_feature_flag.py:456]: Could it be valid to expect this behaviour? And test for it explicitly? Given that you wouldn't be able to get to it anyhow in the UI? Perhaps in the future we handle soft deletes but for now...

Comment (pauldambra) [reply posthog/api/test/test_feature_flag.py:456]: @clarkus what do you think?

This is a case where the API is set-up to GET from `/feature_flags/{the_flag_id}/history` but the library we use to create our routes is helping out when that flag has been deleted and returning 404 before it can get to our history handler

Would we want to have a view in the UI where we can see deleted things? Or are we happy with only seeing history for things that aren't deleted? Do we want to push that into the future?

File: posthog/models/historical_version.py
Comment (hazzadous) [inline posthog/models/historical_version.py:23]: non-blocking

I should also mention, I'm a bit apprehensive about making this table shared across all types. That might be desired for an audit log type query pattern but history for a single object as far as I can see doesn't need a shared table.

Instead, we can have a shared sub-structure on which e.g. `load_history` can operate, whilst keeping data separate.

File: posthog/api/test/test_feature_flag.py
Comment (hazzadous) [inline posthog/api/test/test_feature_flag.py:562]: Can we use something like `"action": "change" | "create" | "delete"` instead which we can use to determine the type of detail e.g. with union types?

We can pull out the key from the detail anyhow, and for history, at least for this change we know that this is FeatureFlag as we're fetching from feature_flags history endpoint

File: posthog/models/history_logging.py
Comment (hazzadous) [inline posthog/models/history_logging.py:187]: Will this result in N+1 queries? Can we add a test that validates that we have O(n) queries.

File: posthog/models/history_logging.py
Comment (hazzadous) [inline posthog/models/history_logging.py:24]: Can we add a bit more detail here type wise? I think we do actually know that we have an `id` and `key`?

File: posthog/models/history_logging.py
Comment (hazzadous) [inline posthog/models/history_logging.py:61]: Can we not just say unknown rather than creating a user in the database?

File: posthog/models/history_logging.py
Comment (hazzadous) [inline posthog/models/history_logging.py:64]: Do we need to add anything to the DB? Isn't this handled in compute_history?

File: posthog/models/history_logging.py
Comment (hazzadous) [reply posthog/models/history_logging.py:24]: Can we also make this an array, one for each key changed? That seems a bit more developer friendly.

File: posthog/models/historical_version.py
Comment (pauldambra) [reply posthog/models/historical_version.py:23]: I'm expecting queries like "changes to insights" but also "changes by user" or "last N changes" e.g. https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=7436%3A40010

What's the source of the apprehension?

File: posthog/api/test/test_feature_flag.py
Comment (pauldambra) [reply posthog/api/test/test_feature_flag.py:562]: hmmmm 🤔 

I'm anticipating the humanizing ending up as Map/Dict lookups

We could split action into a structure `{ type: FeatureFlag, key: filters, action: changed}`

Thinking about [this comment](https://github.com/PostHog/posthog/pull/8931#discussion_r826738883) we might end up with

`{ changes:  [{ type: FeatureFlag, key: filters, action: changed, detail: {stuff: here}}]}`

each item is distinct so could be typed or used as a key in a lookup

File: posthog/models/historical_version.py
Comment (hazzadous) [reply posthog/models/historical_version.py:23]: The apprehension comes from having the schema of `state` owned by different parts of the application, and this table potentially being a hotspot/single point of failure.

I don't have anything more solid than that, I don't want to hold up proceedings.

File: posthog/models/history_logging.py
Comment (pauldambra) [reply posthog/models/history_logging.py]: Turns out it would have done! 😱 

Added `select_related` and a test to protect regression

File: posthog/models/historical_version.py
Comment (pauldambra) [reply posthog/models/historical_version.py:23]: The ones we've thought about are here https://github.com/PostHog/posthog/issues/8545#issuecomment-1035429062

Models that do have a risk of large numbers are event definitions and property definitions (IMO). I'd be in favour of either not tracking their history, or leaving them to much later so we have real data to consider for performance impact

File: posthog/mixins.py
Comment (pauldambra) [reply posthog/mixins.py]: Have rejigged things a bit to keep the existing behaviour and move the destroy method onto the `FeatureFlagViewSet`

File: posthog/api/test/test_feature_flag.py
Comment (pauldambra) [reply posthog/api/test/test_feature_flag.py:517]: You can't (I don't think) automagically associate a migration and a model in code, but you can query migrations by date. So you could write code to show history differently if showing a particular model after the date of a particular migration

I _think_ in your example above we would show the new default field as being changed in the first change after its introduction because it would be in the after state and not the before state

I think the simplest (but most human error-able) solution is that a migration that introduces new state to a model which logs history needs to be followed with a migration that adds a historical version identifying that we have added data to the model

File: posthog/models/historical_version.py
Comment (hazzadous) [reply posthog/models/historical_version.py:92]: Looks like it's an instance, and that you've added a custom serializer to handle this. What is the relation between `instance.__dict__` and `serializer.data`?

Rather than accepting instance here, or serializer for that matter, I would suggest explicitly passing in that you expect for `state` and `name` thereby removing the potential for introducing circular deps (i.e. feature flags now depends on history version, and visa versa)

File: posthog/models/historical_version.py
Comment (hazzadous) [reply posthog/models/historical_version.py:56]: I think we should decide what we intend to be doing here, are we logging, or are we versioning. IMO we should keep versioning and "action" logging separate. We shouldn't be attempting to restore state from a log.

As far as I understand, the immediate requirement is a log.

**Review by hazzadous — CHANGES_REQUESTED** [review]
So I think if this is meant to be an actions log, we should reflect that in the interface:

```
def log_action(organization_id: int, team_id: int, user_id: int, action: str, detail: dict):
    ...
```

If we want this to be versioning, we should just make this a part of feature flags

File: posthog/mixins.py
Comment (hazzadous) [inline posthog/mixins.py:10]: Given that this is what we expect, can we explicitly add to the wrapper signature/typing (and func I guess)?

File: posthog/api/test/test_feature_flag.py
Comment (hazzadous) [inline posthog/api/test/test_feature_flag.py:983]: rather than checking queries explicitly can we:

1. create feature flag as user 1
2. request endpoint and get query count
3. update feature flag as user 2
4. request endpoint and verify counts are the same

this way we're just testing how it scales, not the implementation details of the endpoint

**Review by hazzadous — APPROVED** [review]
Feels much more clear what's going on, and what it's for. Happy to get this in and start getting feedback

Just a couple of comments that you may choose to ignore

**Review by mariusandra — APPROVED** [review]
LGTM. Let's get this going 😅

**Pull request merged**
